File: algorithm_library/python/string/changeable_RollingHash.py
class segtree(): # すべて 0-index
  def __init__(self,V,OP,E):
      self.n=len(V)
      self.op=OP
      self.e=E
      self.log=(self.n-1).bit_length()
      self.size=1<<self.log
      self.data=[E for i in range(2*self.size)]
      for i in range(self.n):
          self.data[self.size+i]=V[i]
      for i in range(self.size-1,0,-1):
          self._update(i)
          
  # 1 点更新
  def set(self,p,x): 
      assert 0<=p and p<self.n
      p+=self.size
      self.data[p]=x
      for i in range(1,self.log+1):
          self._update(p>>i)
          
  # 1 点加算(自分で書いた)
  def add(self,p,x): 
      assert 0<=p and p<self.n
      p+=self.size
      self.data[p] += x
      for i in range(1,self.log+1):
          self._update(p>>i)
          
  # data[p] を返す
  def get(self,p):
      assert 0<=p and p<self.n
      return self.data[p+self.size]
    
  # [l,r) の演算結果を返す
  def prod(self,l,r):
      assert 0<=l and l<=r and r<=self.n
      sml=self.e; smr=self.e
      l+=self.size; r+=self.size
      while(l<r):
          if (l&1):
              sml=self.op(sml,self.data[l])
              l+=1
          if (r&1):
              smr=self.op(self.data[r-1],smr)
              r-=1
          l>>=1
          r>>=1
      return self.op(sml,smr)
    
  # 全区間の演算結果を返す
  def all_prod(self):
      return self.data[1]
    
  # check(operate(data[l],data[l+1],...,data[r-1])) = True
  # を満たす最大の r を返す
  def max_right(self,l,check):
      assert 0<=l and l<=self.n
      assert check(self.e)
      if l==self.n: return self.n
      l+=self.size
      sm=self.e
      while True:
        while(l%2==0):
            l>>=1
        if not(check(self.op(sm,self.data[l]))):
            while(l<self.size):
                l=2*l
                if check(self.op(sm,self.data[l])):
                    sm=self.op(sm,self.data[l])
                    l+=1
            return l-self.size
        sm=self.op(sm,self.data[l])
        l+=1
        if (l&-l)==l: break
      return self.n
  
  # check(operate(data[l],data[l+1],...,data[r-1])) = True
  # を満たす最小の l を返す
  def min_left(self,r,check):
      assert 0<=r and r<=self.n
      assert check(self.e)
      if r==0: return 0
      r+=self.size
      sm=self.e
      while True:
        r-=1
        while(r>1 and (r%2)):
            r>>=1
        if not(check(self.op(self.data[r],sm))):
            while(r<self.size):
                r=(2*r+1)
                if check(self.op(self.data[r],sm)):
                    sm=self.op(self.data[r],sm)
                    r-=1
            return r+1-self.size
        sm=self.op(self.data[r],sm)
        if (r& -r)==r: break
      return 0
    
  def _update(self,k):
      self.data[k]=self.op(self.data[2*k],self.data[2*k+1])
      
  def __str__(self):
      return str([self.get(i) for i in range(self.n)])

class changeable_RollingHash:
  def __init__(self, S, base = 1237, mod = (1<<61)-1):
    N = len(S)
    self.ord_base = ord('a')
    self.base = base
    self.mod = mod 
    self.base_pow = [1] * (N + 1)
    self.base_inv = [1] * (N + 1)
    self.inv = pow(base,mod-2,mod)
    for i in range(N):
      self.base_pow[i+1] = (self.base_pow[i] * base) % mod
      self.base_inv[i+1] = (self.base_inv[i] * self.inv) % mod
    hash = [((ord(s)-self.ord_base+1)*self.base_pow[N-idx-1])%mod for idx,s in enumerate(S)]
    e = 0
    def operate(a,b): return (a+b)%mod
    self.seg = segtree(hash,operate,e)

  # インデックス x の要素を c に変える
  def change(self, x, c): 
    self.seg.set(x,((ord(c)-self.ord_base+1)*self.base_pow[self.N-x-1])%self.mod)
    
  def get_hash(self, l, r): #[l, r),0-index
    return (self.seg.prod(l,r) * self.base_inv[self.N-r]) % self.mod

File: algorithm_library/cpp-from-py/string/changeable_RollingHash.cpp
#include <bits/stdc++.h>
using namespace std;

struct SumSegtree {
    int n, size, log;
    unsigned long long mod;
    vector<unsigned long long> data;

    SumSegtree() {}
    SumSegtree(const vector<unsigned long long>& V, unsigned long long mod_) : n(V.size()), mod(mod_) {
        log = 0;
        size = 1;
        while (size < n) size <<= 1, log++;
        data.assign(2 * size, 0);
        for (int i = 0; i < n; i++) data[size + i] = V[i] % mod;
        for (int i = size - 1; i > 0; i--) update(i);
    }

    void update(int k) { data[k] = (data[2 * k] + data[2 * k + 1]) % mod; }
    void set_val(int p, unsigned long long x) {
        assert(0 <= p && p < n);
        p += size;
        data[p] = x % mod;
        for (int i = 1; i <= log; i++) update(p >> i);
    }
    unsigned long long prod(int l, int r) const {
        assert(0 <= l && l <= r && r <= n);
        unsigned long long sml = 0, smr = 0;
        for (l += size, r += size; l < r; l >>= 1, r >>= 1) {
            if (l & 1) sml = (sml + data[l++]) % mod;
            if (r & 1) smr = (data[--r] + smr) % mod;
        }
        return (sml + smr) % mod;
    }
};

struct changeable_RollingHash {
    int N;
    int ord_base = 'a';
    unsigned long long base, mod, inv;
    vector<unsigned long long> base_pow, base_inv;
    SumSegtree seg;

    changeable_RollingHash(const string& S, unsigned long long base_ = 1237, unsigned long long mod_ = (1ULL << 61) - 1)
        : N(S.size()), base(base_), mod(mod_) {
        inv = mod_pow(base, mod - 2);
        base_pow.assign(N + 1, 1);
        base_inv.assign(N + 1, 1);
        for (int i = 0; i < N; i++) {
            base_pow[i + 1] = mul(base_pow[i], base);
            base_inv[i + 1] = mul(base_inv[i], inv);
        }
        vector<unsigned long long> h(N);
        for (int idx = 0; idx < N; idx++) h[idx] = mul(S[idx] - ord_base + 1, base_pow[N - idx - 1]);
        seg = SumSegtree(h, mod);
    }

    unsigned long long mul(unsigned long long a, unsigned long long b) const {
        return (__uint128_t)a * b % mod;
    }

    unsigned long long mod_pow(unsigned long long a, unsigned long long e) const {
        unsigned long long r = 1;
        while (e) {
            if (e & 1) r = (__uint128_t)r * a % mod;
            a = (__uint128_t)a * a % mod;
            e >>= 1;
        }
        return r;
    }

    void change(int x, char c) {
        seg.set_val(x, mul(c - ord_base + 1, base_pow[N - x - 1]));
    }

    unsigned long long get_hash(int l, int r) const {
        return mul(seg.prod(l, r), base_inv[N - r]);
    }
};
